File: Go/work/src/benchmark/fib/fib.go
// fib.go
// 
// vim: ft=go sw=4 ts=4
//
// Calculating Fibonacci numbers in Go
//

package fib

import "math/big"

// naive recursive implementation
func FibNaive(n int) int {
	if (n < 2) {
		return n
	} else {
		return FibNaive(n-1) + FibNaive(n-2)
	}
}

// tco optimized version
func fibAux(n int, a, b *big.Int) *big.Int {
	if (n == 0) {
		return a
	} else {
		return fibAux(n-1, b, a.Add(a, b))
	}
}

func Fib(n int) *big.Int {
	return fibAux(n, big.NewInt(0), big.NewInt(1))
}

File: JS/Benchmark.ts/fib.ts
// fib.js
// Calculate Fibonacci numbers in JavaScript
//
// vim: ft=javascript sw=4 ts=4
//

//var BigInt = require('@stdlib/bigint');

// naive & efficient (BigInt) implementation
export function fib_naive(n:number): number {
	return (n < 2) ? n : fib_naive(n - 1) + fib_naive(n - 2);
}

function _fib_aux(n:number, a:bigint, b:bigint): bigint {
	return (n === 0) ? a : _fib_aux(n - 1, b, a + b );
}

export function fib(n:number): bigint {
	return _fib_aux(n, BigInt(0), BigInt(1));
}
